File: 2642. Design Graph With Shortest Path Calculator/Solution.py
class Graph:

    def __init__(self, n: int, edges: List[List[int]]):
        self.d = [[2000000000 for i in range(n)] for j in range(n)]
        self.n = n
        for i in range(n): self.d[i][i] = 0
        for x, y, z in edges: self.d[x][y] = z
        self.update()
    
    def update(self):
        for k in range(self.n):
            for i in range(self.n):
                for j in range(self.n):
                    self.d[i][j] = min(self.d[i][j], self.d[i][k] + self.d[k][j])

    def addEdge(self, edge: List[int]) -> None:
        self.d[edge[0]][edge[1]] = edge[2]

    def shortestPath(self, node1: int, node2: int) -> int:
        return self.d[node1][node2] if self.d[node1][node2] != 2000000000 else -1


# Your Graph object will be instantiated and called as such:
# obj = Graph(n, edges)
# obj.addEdge(edge)
# param_2 = obj.shortestPath(node1,node2)

File: 2642. Design Graph With Shortest Path Calculator/Solution.java
class Graph {

    public long[][] d;
    public int n;
    
    public Graph(int n, int[][] edges) {
        d = new long[n][n];
        this.n = n;
        for (int i = 0; i < n; i++) {
            Arrays.fill(d[i], 2000000000);
            d[i][i] = 0;
        }
        for (int[] e : edges) d[e[0]][e[1]] = e[2];
        update();
    }
    
    public void addEdge(int[] edge) {
        d[edge[0]][edge[1]] = edge[2];
        update();
    }
    
    public void update() {
        for (int k = 0; k < n; k++) {
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n; j++) {
                    d[i][j] = Math.min(d[i][j], d[i][k] + d[k][j]);
                }
            }
        }
    }
    
    public int shortestPath(int node1, int node2) {
        return (d[node1][node2] == 2000000000) ? -1 : (int) d[node1][node2];
    }
}

/**
 * Your Graph object will be instantiated and called as such:
 * Graph obj = new Graph(n, edges);
 * obj.addEdge(edge);
 * int param_2 = obj.shortestPath(node1,node2);
 */